<launch>
  <!-- 单立方体调试：相机 + 棱方体检测 + relay + world_tag + 单机融合 + debug -->

  <arg name="robot_name" default="jaka1" />
  <arg name="robot_id" default="$(eval {'jaka1':1,'jaka2':2,'jaka3':3,'jaka4':4}.get(arg('robot_name'),1))" />
  <arg name="robot_ids" default="$(eval '[' + str(arg('robot_id')) + ']')" />
  <arg name="faces_yaml" default="$(find jaka_close_contro)/config/cube_faces_4robots.yaml" />
  <arg name="output_csv" default="$(find jaka_close_contro)/config/cube_fusion_debug.csv" />
  <arg name="stats_window" default="0" />
  <arg name="print_rate" default="10.0" />
  <arg name="fused_from_tf" default="false" />
  <arg name="camera_frame" default="" />
  <arg name="fused_frame" default="cube_center_$(arg robot_name)" />
  <arg name="svo_file" default="" />

  <!-- 1. ZED2i 相机 -->
  <include file="$(find zed_wrapper)/launch/zed2i.launch">
    <arg name="svo_file" value="$(arg svo_file)" />
  </include>

  <!-- 2. ArUco 检测与分流 -->
  <node pkg="jaka_close_contro" type="cube_aruco_detector_node" name="cube_aruco_detector" output="screen">
    <param name="image_topic"       value="/zed2i/zed_node/left/image_rect_color" />
    <param name="camera_info_topic" value="/zed2i/zed_node/left/camera_info" />
    <param name="fiducial_topic"    value="/fiducial_transforms" />
    <param name="default_marker_size" value="0.08" />
    <param name="dictionary" value="DICT_4X4_50" />
    <param name="tool_dictionary" value="DICT_6X6_1000" />
    <param name="tool_marker_length_m" value="0.0425" />
    <param name="tool_marker_separation_m" value="0.005" />
    <param name="min_gridboard_markers" value="2" />
    <param name="world_board_enable" value="true" />
    <param name="world_board_output_id" value="500" />
    <param name="world_board_marker_length_m" value="0.0525" />
    <param name="world_board_marker_separation_m" value="0.005" />
    <param name="world_board_min_markers" value="2" />
    <param name="enable_single_world_markers" value="false" />
    <param name="robot_stride" value="100" />
    <param name="id_step" value="10" />
    <param name="face_id_base" value="10" />
    <rosparam subst_value="true" param="robot_ids">$(arg robot_ids)</rosparam>
    <rosparam param="marker_sizes">
      "0": 0.08
    </rosparam>
    <rosparam param="world_board_marker_ids">[500, 501, 502, 503]</rosparam>
  </node>

  <node pkg="jaka_close_contro" type="fiducial_relay_node" name="fiducial_relay" output="screen">
    <rosparam param="world_ids">[500, 501, 502, 503]</rosparam>
    <rosparam subst_value="true" param="robot_ids">$(arg robot_ids)</rosparam>
    <param name="input_topic" value="/fiducial_transforms" />
    <param name="output_world_topic" value="/world_fiducials" />
    <param name="output_tool_topic" value="/tool_fiducials" />
  </node>

  <!-- 3. 世界坐标系定义 -->
  <node pkg="jaka_close_contro" type="world_tag_node" name="world_tag_node" output="screen">
    <param name="world_id" type="int" value="500" />
    <rosparam param="world_ids">[500, 501, 502, 503]</rosparam>
    <param name="fiducial_topic"    value="/world_fiducials" />
    <param name="world_frame"       value="world" />
    <param name="camera_frame"      value="zed2i_left_camera_optical_frame" />
    <param name="alpha_pos"         value="0.3" />
    <param name="alpha_rot"         value="0.3" />
  </node>

  <!-- 4. 单机器人融合 -->
  <node pkg="jaka_close_contro" type="cube_multi_face_fusion_node" name="cube_multi_face_fusion_single" output="screen">
    <param name="faces_yaml"           value="$(arg faces_yaml)" />
    <param name="fiducial_topic"       value="/tool_fiducials" />
    <param name="cube_frame"           value="cube_center_$(arg robot_name)" />
    <param name="camera_frame_default" value="zed2i_left_camera_optical_frame" />
    <param name="timeout_sec"          value="0.5" />
    <param name="robot_id" value="$(arg robot_id)" />
    <param name="robot_name" value="$(arg robot_name)" />
    <param name="face_id_base" value="10" />
    <param name="robot_stride" value="100" />
    <param name="output_topic" value="/vision/$(arg robot_name)/cube_center" />
    <param name="stats_topic" value="/vision/$(arg robot_name)/cube_fusion_stats" />
    <param name="filter_robot_faces" value="true" />
    <param name="broadcast_tf" value="true" />
  </node>

  <!-- 5. Debug：仅关注选中立方体 -->
  <node pkg="jaka_close_contro" type="cube_fusion_debug_node" name="cube_fusion_debug_single" output="screen">
    <param name="faces_yaml"      value="$(arg faces_yaml)" />
    <param name="fiducial_topic"  value="/tool_fiducials" />
    <param name="fused_pose_topic" value="/vision/$(arg robot_name)/cube_center" />
    <param name="camera_frame" value="$(arg camera_frame)" />
    <param name="fused_frame" value="$(arg fused_frame)" />
    <param name="fused_from_tf" value="$(arg fused_from_tf)" />
    <param name="csv_path"      value="$(arg output_csv)" />
    <param name="print_rate"    value="$(arg print_rate)" />
    <param name="stats_window"  value="$(arg stats_window)" />
    <param name="filter_robot_faces" value="true" />
    <param name="robot_id" value="$(arg robot_id)" />
    <param name="face_id_base" value="10" />
    <param name="robot_stride" value="100" />
    <param name="num_faces" value="4" />
    <param name="tf_prefix" value="$(arg robot_name)_" />
  </node>
</launch>
